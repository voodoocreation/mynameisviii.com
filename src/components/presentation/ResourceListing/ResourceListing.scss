@import "../../../scss/variables", "../../../scss/mixins";

.ResourceListing {
  flex-basis: 100%;
  padding: $spacing-half;

  h3 {
    margin-top: 0;
  }

  p {
    margin-bottom: 0;
  }

  a {
    @include panel-2();

    display: flex;
    flex-direction: column;
    height: 100%;
    border-radius: $radius;
    transition: background-color 200ms ease-out;

    &:hover {
      background: $panel-base-color;
    }
  }

  img {
    margin: auto;
  }

  &--image {
    display: flex;
    flex-basis: 100%;
    flex-shrink: 0;
    align-items: center;
    padding: $spacing;

    .Loader {
      padding-top: 100%;
    }
  }

  &--details {
    flex-grow: 1;
    padding: 0 $spacing $spacing;
  }

  @include screen-above($screen-xs-min) {
    a {
      flex-direction: row;
    }

    &--image {
      flex-basis: 33.333333%;
      padding: $spacing $spacing-half $spacing $spacing;
    }

    &--details {
      padding: $spacing $spacing $spacing $spacing-half;
    }
  }

  @include screen-above(590px) {
    &--image {
      flex-basis: 25%;
    }
  }

  @include screen-above($screen-lg-min) {
    flex-basis: 50%;

    &--image {
      flex-basis: 33.333333%;
    }
  }

  @include screen-above(1500px) {
    flex-basis: 33.333333%;
  }
}
