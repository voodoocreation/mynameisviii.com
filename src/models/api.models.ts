import { arrayToAssoc } from "../transformers/transformData";

interface IWithSlug {
  slug: string;
}

export type TLastEvaluatedKey<K extends string> = Record<
  K | "isActive" | "slug",
  string
>;

export interface IRawDynamoResponse<T extends {}, K extends string> {
  items: T[];
  lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IDynamoResponse<T extends IWithSlug, K extends string> {
  readonly items: Record<string, T>;
  readonly lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IRawS3Response<T extends {}> {
  isTruncated?: boolean;
  items: T[];
}

export interface IS3Response<T extends IWithSlug> {
  readonly isTruncated: boolean;
  readonly items: Record<string, T>;
}

export const dynamoResponse = <T extends IWithSlug, K extends string>(
  options: IRawDynamoResponse<T, K>
): IDynamoResponse<T, K> => ({
  items: arrayToAssoc(options.items, "slug"),
  lastEvaluatedKey: options.lastEvaluatedKey
});

export const s3Response = <T extends IWithSlug>(
  options: IRawS3Response<T>
): IS3Response<T> => ({
  isTruncated: !!options.isTruncated,
  items: arrayToAssoc(options.items, "slug")
});
