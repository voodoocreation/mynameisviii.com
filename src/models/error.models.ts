export interface IError {
  readonly message: string;
  readonly status: number;
}

export const error = (options: Partial<IError> = {}): IError => ({
  message: options.message || "An error has occurred",
  status: options.status || 500,
});
