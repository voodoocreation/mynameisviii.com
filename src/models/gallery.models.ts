import dayjs from "dayjs";

export interface IRawGallery {
  description?: string;
  imageUrl?: string;
  images?: Array<Partial<IGalleryImage>>;
  modifiedAt?: string;
  slug?: string;
  title?: string;
}

export interface IGallery {
  readonly description?: string;
  readonly imageUrl: string;
  readonly images?: IGalleryImage[];
  readonly modifiedAt: string;
  readonly slug: string;
  readonly title: string;
}

export interface IGalleryImage {
  readonly imageUrl: string;
  readonly modifiedAt: string;
}

export const galleryImage = (
  options: Partial<IGalleryImage> = {}
): IGalleryImage => ({
  imageUrl: options.imageUrl || "",
  modifiedAt: options.modifiedAt || dayjs().toISOString()
});

export const gallery = (options: IRawGallery = {}): IGallery => ({
  description: options.description,
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(galleryImage) : undefined,
  modifiedAt: options.modifiedAt || dayjs().toISOString(),
  slug: options.slug || "",
  title: options.title || ""
});
