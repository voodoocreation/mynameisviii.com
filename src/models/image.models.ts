export interface IImage {
  readonly imageUrl: string;
  readonly title: string;
}

export const image = (options: Partial<IImage> = {}): IImage => ({
  imageUrl: options.imageUrl || "",
  title: options.title || ""
});
