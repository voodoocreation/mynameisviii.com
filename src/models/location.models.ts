import { TYPE } from "../constants/location.constants";

export interface ILatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface ILocation {
  readonly address?: string;
  readonly latLng?: ILatLng;
  readonly name: string;
  readonly type: TYPE;
  readonly url?: string;
}

export const location = (options: Partial<ILocation> = {}): ILocation => ({
  address: options.address,
  latLng: options.latLng,
  name: options.name || "",
  type: options.type || TYPE.MUSIC_VENUE,
  url: options.url,
});
