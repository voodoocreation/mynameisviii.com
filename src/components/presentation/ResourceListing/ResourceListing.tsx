import cn from "classnames";
import * as React from "react";

import { IResource } from "../../../models/root.models";
import Image from "../Image/Image";
import Link from "../Link/Link";

import "./ResourceListing.scss";

interface IProps extends IResource {
  onLoad: () => void;
}

interface IState {
  isRendered: boolean;
}

class ResourceListing extends React.Component<IProps, IState> {
  public state = {
    isRendered: false
  };

  public render() {
    const { onLoad, ...resource } = this.props;
    const { isRendered } = this.state;

    return (
      <article className={cn("ResourceListing", { isRendered })}>
        <Link href={resource.url} isExternal>
          <Image
            alt={resource.title}
            className="ResourceListing--image"
            src={resource.imageUrl}
            onLoad={this.onLoad}
          />

          <div className="ResourceListing--details">
            <h3>{resource.title}</h3>
            <p>{resource.description}</p>
          </div>
        </Link>
      </article>
    );
  }

  private onLoad = () => {
    this.setState({
      isRendered: true
    });

    if (this.props.onLoad) {
      this.props.onLoad();
    }
  };
}

export default ResourceListing;
