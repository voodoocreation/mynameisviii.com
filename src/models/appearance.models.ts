import dayjs from "dayjs";

import { BOOLEAN } from "../constants/api.constants";
import { STATUS, TYPE } from "../constants/appearance.constants";
import { IImage, image } from "./image.models";
import { ILocation, location } from "./location.models";
import { IOffer, offer } from "./offer.models";
import { IOrganization, organization } from "./organization.models";
import { IPerformer, IRawPerformer, performer } from "./performer.models";

export interface IRawAppearance {
  acts?: Array<Partial<IRawPerformer>>;
  audience?: string;
  description?: string;
  finishingAt?: string;
  imageUrl?: string;
  images?: IImage[];
  isActive?: BOOLEAN;
  location?: Partial<ILocation>;
  ogImageUrl?: string;
  organizer?: Partial<IOrganization>;
  rsvpUrl?: string;
  sales?: Array<Partial<IOffer>>;
  slug?: string;
  startingAt?: string;
  status?: STATUS;
  title?: string;
  type?: TYPE;
}

export interface IAppearance {
  readonly acts: IPerformer[];
  readonly audience?: string;
  readonly description: string;
  readonly finishingAt: string;
  readonly imageUrl: string;
  readonly images: IImage[];
  readonly isActive: boolean;
  readonly location: ILocation;
  readonly ogImageUrl: string;
  readonly organizer: IOrganization;
  readonly rsvpUrl?: string;
  readonly sales: IOffer[];
  readonly slug: string;
  readonly startingAt: string;
  readonly status: STATUS;
  readonly title: string;
  readonly type: TYPE;
}

export const appearance = (options: IRawAppearance = {}): IAppearance => ({
  acts: options.acts ? options.acts.map(performer) : [],
  audience: options.audience,
  description: options.description || "",
  finishingAt: options.finishingAt || dayjs().toISOString(),
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(image) : [],
  isActive: (options.isActive && options.isActive === BOOLEAN.TRUE) || false,
  location: location(options.location),
  ogImageUrl: options.ogImageUrl || "",
  organizer: organization(options.organizer),
  rsvpUrl: options.rsvpUrl,
  sales: options.sales ? options.sales.map(offer) : [],
  slug: options.slug || "",
  startingAt: options.startingAt || dayjs().toISOString(),
  status: options.status || STATUS.SCHEDULED,
  title: options.title || "",
  type: options.type || TYPE.MUSIC,
});
