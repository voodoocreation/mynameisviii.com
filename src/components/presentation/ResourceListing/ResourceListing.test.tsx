import { resource } from "../../../models/root.models";
import ComponentTester from "../../../utilities/ComponentTester";
import ResourceListing from "./ResourceListing";

const component = new ComponentTester(ResourceListing).withDefaultProps(
  resource({
    createdAt: "2017-10-10T18:00:00",
    description: "Description",
    imageUrl: "Image URL",
    slug: "resource-1",
    title: "Title",
    url: "URL"
  })
);

describe("[presentation] <ResourceListing />", () => {
  const { props, wrapper } = component
    .withProps({
      onLoad: jest.fn()
    })
    .mount();

  it("doesn't add the isRendered class initially", () => {
    expect(wrapper.find(".isRendered")).toHaveLength(0);
  });

  it("loads the image", () => {
    wrapper.find("Image.ResourceListing--image").simulate("load");
  });

  it("adds the isRendered class", () => {
    expect(wrapper.find(".isRendered")).toHaveLength(1);
  });

  it("calls the onLoad prop", () => {
    expect(props.onLoad).toHaveBeenCalledTimes(1);
  });

  it("sets the onLoad prop to be undefined", () => {
    wrapper.setProps({ onLoad: undefined });
  });

  it("loads the image", () => {
    wrapper.find("Image.ResourceListing--image").simulate("load");
  });

  it("matches snapshot", () => {
    expect(wrapper.render()).toMatchSnapshot();
  });
});
