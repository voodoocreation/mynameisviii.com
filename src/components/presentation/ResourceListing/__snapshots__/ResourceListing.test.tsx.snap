// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <ResourceListing /> matches snapshot 1`] = `
<article
  class="ResourceListing isRendered"
>
  <a
    href="URL"
    rel="noopener"
    target="_blank"
  >
    <figure
      class="Image ResourceListing--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
    <div
      class="ResourceListing--details"
    >
      <h3>
        Title
      </h3>
      <p>
        Description
      </p>
    </div>
  </a>
</article>
`;
